### PR TITLE
Jordan's branch

### PR DESCRIPTION
**Pull request by Eric-Marchetti**

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 1]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
...
+            startShooter();
+        }
+
+        else if (oi.autoShootButton()) {
+            if (checkVision()) {
+                shooterState = ShooterState.AUTOMATIC;
+            }
+
+            else {
+                shooterState = ShooterState.MANUAL;
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: You need to maintain the last reading not switch to a separate level if there is a miss on the vision

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 2]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
         shooter.setNeutralMode(NeutralMode.Coast);
         shooter.configSupplyCurrentLimit(shooterLimit);
 
+        ml.createfile("shooterRPMS");
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: We do not need to replay this from a file

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 3]
```diff
@@ -34,16 +39,18 @@
...
 
     // States for gatherer
-
     public static final int gathererUp = 0;
     public static final int gathererDown = 1;
     public static final int holding = 3;
 
-    public int state = gathererUp;
+    public int gathererState = gathererUp;
+    public ShooterState shooterState = ShooterState.MANUAL;
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: It doesn't seem like this is needed

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 4]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
...
-            lowerIntake();
-            startIntake();
-        } else if (!oi.manualIntakeButton() && state == gathererDown) {
-            stopIntake();
-        } else if (oi.manualIntakeButton() && state == holding) {
-            stopIntake();
-            raiseIntake();
+        switch (gathererState) {
+            case gathererUp:
+                if (oi.manualIntakeButton()) {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: needs else: stop intake

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 5]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
...
+                    lowerIntake();
+                    startIntake();
+                }
+                break;
+            case gathererDown:
+                if (!oi.manualIntakeButton()) {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: needs else: run intake

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 6]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
...
+                    stopIntake();
+                }
+                break;
+            case holding:
+                if (oi.manualIntakeButton()) {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: needs else: stop intake

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 7]
```diff
@@ -76,54 +83,123 @@ public Shooter(OperatorInterface operatorinterface) {
...
     public void runShooter() {
         // System.out.println(shooter.getSelectedSensorVelocity()); PRINTS FOR VELOCITY
         // CONTROL
 
         // Control for FlyWheel
+
         if (oi.runFlyWheelButtonManual()) {
-            startShooter(shooterRpms);
-        } else {
+            shooterState = ShooterState.MANUAL;
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: remove shooter state variable, overcomplicates things

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 8]
```diff
@@ -1,6 +1,8 @@
 package frc.robot.subsystems;
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java:1]: Most method names need to be rethought/combined in order to make this readable.

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 9]
```diff
@@ -132,8 +208,38 @@ public void startIntake() {
...
+
+        if (shooterRpms < 0) {
+            shooterRpms = 0;
+        }
+
+        else if (shooterRpms > 15000) {
+            shooterRpms = 15000;
+        }
+
+        ml.write(Double.toString(shooterRpms));
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: Don't need this recorded anymore

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 10]
```diff
@@ -132,8 +208,38 @@ public void startIntake() {
...
+        }
+        return false;
+    }
+
+    public void autoShoot() {
+        vc.autoShoot();
+    }
+
+    private void updateManualRPMS() {
+        if (oi.copilot.getDPadPress(oi.DPadUp)) {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java:230]: ratelimit

File: src/main/java/frc/robot/subsystems/VisionControl.java [hunk 11]
```diff
@@ -6,19 +6,26 @@
 
 @SuppressWarnings("unused")
 public class VisionControl {
+    public enum autoState {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/VisionControl.java:12]: Auto Code really needs to be in a separate PR so that we can finish and merge this

File: src/main/java/frc/robot/subsystems/VisionControl.java [hunk 12]
```diff
@@ -272,17 +265,48 @@ private double calculateBallRotation() {
...
+                }
+                break;
+        }
+    }
+
+    public boolean isHoopValid() {
+        return visionData.isHoopValid();
+    }
+
+    private double calculateShooterRPMS() {
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/VisionControl.java]: We need to fix this once we test and get the calibration curve

File: src/main/java/frc/robot/subsystems/VisionControl.java [hunk 13]
```diff
@@ -272,17 +265,48 @@ private double calculateBallRotation() {
...
+            case WAIT:
+                double rpm = calculateShooterRPMS();
+                shooter.startShoter(rpm);
+                if (Math.abs(shooter.getShooterRpms() - rpm) < shooterThreshold) {
+                    shooterstate = shooterState.SHOOT;
+                }
+                break;
+            case SHOOT:
+                double rpms = calculateShooterRPMS();
+                shooter.startShoter(rpms);
```
Comment (Eric-Marchetti) [inline src/main/java/frc/robot/subsystems/VisionControl.java]: Can just set the feed motor power for a time then be done

File: src/main/java/frc/robot/subsystems/VisionControl.java (hunk 12 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/VisionControl.java]: yep

File: src/main/java/frc/robot/subsystems/VisionControl.java (hunk 11 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/VisionControl.java:12]: yep

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 10 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java:230]: already fixed

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 9 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: yes we do

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 2 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: not bein recorded

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 1 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: resolved

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 3 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: shooter states are now gone

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 7 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: already done

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 8 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java:1]: done

File: src/main/java/frc/robot/subsystems/VisionControl.java (hunk 13 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/VisionControl.java]: done

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 4 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: done

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 5 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: done

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 6 above)
Comment (Rylan-Meilutis) [reply src/main/java/frc/robot/subsystems/Shooter.java]: done

File: src/main/java/frc/robot/FeatureFlags.java [hunk 14]
```diff
@@ -4,17 +4,17 @@
 public final class FeatureFlags {
 
     // Modify these values
-    public static boolean doVision = true;
-    public static boolean doChassis = true;
-    public static boolean doImu = true;
-    public static boolean doShooter = false;
+    public static boolean doVision = false;
+    public static boolean doChassis = false;
```
Comment (mmarchetti) [inline src/main/java/frc/robot/FeatureFlags.java]: remember to make this true again

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 15]
```diff
@@ -34,106 +36,238 @@
...
+            }
+        }
+    }
+
+    public void ShooterMain() {
+        System.out.println(shooterRpms);
+        updateManualRPMS();
+
         if (oi.runFlyWheelButtonManual()) {
+            if (checkVision()) {
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: we should not check vision here, just run the shooter in manual mode

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 16]
```diff
@@ -34,106 +36,238 @@
...
-        } else {
+        }
+
+        else if (oi.autoShootButton()) {
+            if (checkVision()) {
+                autoShoot();
+            }
+
+            else {
+                if (checkVision()) {
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: we just called checkVision above, so this call is not needed

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 17]
```diff
@@ -34,106 +36,238 @@
...
+                    vc.shooterstate = shooterState.ALIGN;
+                }
+                startShooter(shooterRpms);
+                feederMain();
+            }
+        }
+
+        else {
+            if (checkVision()) {
+                vc.shooterstate = shooterState.ALIGN;
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: not sure what this does

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 18]
```diff
@@ -34,106 +36,238 @@
...
-        } else if (!oi.manualIntakeButton() && state == gathererDown) {
-            stopIntake();
-        } else if (oi.manualIntakeButton() && state == holding) {
-            stopIntake();
-            raiseIntake();
+    public void feederMain() {
+        if (!oi.autoShootButton()) {
+            if (oi.shootButton()) {
+                startFeeder(feederSpeed);
+                startIntake(-intakeSpeed);
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: why is this negative?

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 19]
```diff
@@ -34,106 +36,238 @@
...
-    public void startIntake() {
-        intake.set(TalonSRXControlMode.PercentOutput, intakeSpeed);
+    public void startIntake(double speed) {
+        intake.set(TalonSRXControlMode.PercentOutput, speed);
     }
 
     public void stopIntake() {
-        state = holding;
         intake.set(TalonSRXControlMode.PercentOutput, 0);
+        gathererState = holding;
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java]: This probably isnt' the right place to change to holding state, since we stop the intake in different states

File: src/main/java/frc/robot/subsystems/Shooter.java [hunk 20]
```diff
@@ -34,106 +36,238 @@
...
+        feederMain();
+
+        // Control for lowering intake
+        gathererMain();
+    }
+
+    public void gathererMain() {
+        if (FeatureFlags.doCompressor && FeatureFlags.compressorInitialized) {
+            switch (gathererState) {
+                case gathererUp:
```
Comment (mmarchetti) [inline src/main/java/frc/robot/subsystems/Shooter.java:118]: do the state changes in this function instead of the other functions

**Review by mmarchetti — APPROVED** [review]

File: src/main/java/frc/robot/subsystems/Shooter.java (hunk 17 above)
Comment (pyjshtml) [reply src/main/java/frc/robot/subsystems/Shooter.java:175]: ShooterMain needs to be completely remade
